src/sentry/services/hybrid_cloud/rpc.py

```diff
@@ -0,0 +1,270 @@
+
+    def deserialize_rpc_arguments(
+        self, method_name: str, serial_arguments: ArgumentDict
+    ) -> pydantic.BaseModel:
+        signature = self._signatures[method_name]
+        return signature.deserialize_arguments(serial_arguments)
+
+
+def rpc_method(method: Callable[..., _T]) -> Callable[..., _T]:
+    if not getattr(method, "__isabstractmethod__", False):
+        raise RpcServiceSetupException("`@rpc_method` may only decorate abstract methods")
+    setattr(method, _IS_RPC_METHOD_ATTR, True)
+    return method
+
+
+_global_service_registry: Dict[str, DelegatingRpcService] = {}
+
+
+class RpcService(InterfaceWithLifecycle):
+    name: str
+    local_mode: SiloMode
+
+    _signatures: Mapping[str, RpcMethodSignature]
+
+    def __init_subclass__(cls) -> None:
+        if cls._declares_service_interface():
+            # These class attributes are required on any RpcService subclass that has
+            # at least one method decorated by `@rpc_method`. (They can be left off
+            # if and when we make an intermediate abstract class.)
+            if not isinstance(getattr(cls, "name", None), str):
+                raise RpcServiceSetupException("`name` class attribute (str) is required")
+            if not isinstance(getattr(cls, "local_mode", None), SiloMode):
+                raise RpcServiceSetupException(
+                    "`local_mode` class attribute (SiloMode) is required"
+                )
+        cls._signatures = cls._create_signatures()
+
+    @classmethod
+    def _get_all_abstract_methods(cls) -> Iterator[Callable[..., Any]]:
+        for attr_name in dir(cls):
+            attr = getattr(cls, attr_name, None)
+            if callable(attr) and getattr(attr, "__isabstractmethod__", False):
+                yield attr
+
+    @classmethod
+    def _get_all_abstract_rpc_methods(cls) -> Iterator[Callable[..., Any]]:
+        return (
+            m for m in cls._get_all_abstract_methods() if getattr(m, _IS_RPC_METHOD_ATTR, False)
+        )
+
+    @classmethod
+    def _declares_service_interface(cls) -> bool:
+        """Check whether a subclass declares the service interface.
+
+        By "service interface", we mean the set of RPC methods that are offered.
+        Those methods are decorated by `@rpc_method`, so we return true for any
+        service class that declares (but does not implement) at least one such method.
+
+        This method would return false, for example, for the local, database-backed
+        implementation that inherits from the base service. It also would return
+        false on an intermediate, abstract subclass that is meant to be extended to
+        declare other base services.
+        """
+
+        try:
+            next(cls._get_all_abstract_rpc_methods())
+        except StopIteration:
+            return False
+        else:
+            return True
+
+    @classmethod
+    @abstractmethod
+    def get_local_implementation(cls) -> RpcService:
```

Comment:
This is a good improvement over how local/remote delegation was wired up before :clap: